File: python/Q212_word_search_ii.py
#
# @lc app=leetcode id=212 lang=python3
#
# [212] Word Search II
#


# @lc code=start
class TrieNode:
    def __init__(self) -> None:
        self.children: dict[str, TrieNode] = dict()
        self.end: bool = False


class Solution:
    def insert(self, word: str) -> None:
        current: TrieNode = self.words

        for c in word:
            if c not in current.children:
                current.children[c] = TrieNode()

            current = current.children[c]

        current.end = True

    def find(self, row: int, col: int, node: TrieNode, word: list[str]) -> None:
        if node.end:
            self.result.add("".join(word))

        # out of bounds, visited cell, or no valid words left
        if (
            row < 0
            or col < 0
            or row >= self.ROWS
            or col >= self.COLS
            or self.visiting[row][col]
            or self.board[row][col] not in node.children
        ):
            return

        next_node: TrieNode = node.children[self.board[row][col]]

        self.visiting[row][col] = True
        word.append(self.board[row][col])
        for d1, d2 in self.directions:
            self.find(row + d1, col + d2, next_node, word)
        word.pop()
        self.visiting[row][col] = False

    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        self.ROWS: int = len(board)
        self.COLS: int = len(board[0])

        self.board: list[list[str]] = board
        self.directions: list[tuple[int, int]] = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        self.visiting: list[list[bool]] = [
            [False for _ in range(self.COLS)] for _ in range(self.ROWS)
        ]

        self.result: set[str] = set()

        self.words: TrieNode = TrieNode()
        for word in words:
            self.insert(word)

        for i in range(self.ROWS):
            for j in range(self.COLS):
                self.find(i, j, self.words, [])

        return list(self.result)


# @lc code=end

File: rust/src/leetcode/Q212_word_search_ii.rs
/*
 * @lc app=leetcode id=212 lang=rust
 *
 * [212] Word Search II
 */

// @lc code=start
use std::collections::HashMap;
use std::collections::HashSet;

struct TrieNode {
    children: HashMap<char, Box<TrieNode>>,
    word: bool,
}

impl TrieNode {
    #[inline]
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            word: false,
        }
    }
}

impl Solution {
    fn dictionary(words: Vec<String>, node: &mut TrieNode) {
        for word in words {
            let mut current: &mut TrieNode = node;

            for c in word.chars() {
                current = current
                    .children
                    .entry(c)
                    .or_insert(Box::new(TrieNode::new()));
            }

            current.word = true;
        }
    }

    fn dfs(
        row: i32,
        col: i32,
        board: &Vec<Vec<char>>,
        visited: &mut HashSet<(i32, i32)>,
        current: &mut Vec<char>,
        node: &TrieNode,
        result: &mut HashSet<String>,
    ) {
        if row < 0
            || col < 0
            || row >= board.len() as i32
            || col >= board[0].len() as i32
            || visited.contains(&(row, col))
        {
            return;
        }

        if let Some(next_node) = node.children.get(&board[row as usize][col as usize]) {
            visited.insert((row, col));
            current.push(board[row as usize][col as usize]);

            if next_node.word {
                result.insert(current.iter().collect());
            }

            Self::dfs(
                row + 1,
                col,
                board,
                visited,
                current,
                next_node,
                result,
            );
            Self::dfs(
                row - 1,
                col,
                board,
                visited,
                current,
                next_node,
                result,
            );
            Self::dfs(
                row,
                col + 1,
                board,
                visited,
                current,
                next_node,
                result,
            );
            Self::dfs(
                row,
                col - 1,
                board,
                visited,
                current,
                next_node,
                result,
            );

            visited.remove(&(row, col));
            current.pop();
        }
    }

    pub fn find_words(board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        let mut root = TrieNode::new();
        Self::dictionary(words, &mut root);

        let mut current: Vec<char> = vec![];
        let mut visited: HashSet<(i32, i32)> = HashSet::new();

        let mut result: HashSet<String> = HashSet::new();

        for i in 0..board.len() {
            for j in 0..board[0].len() {
                Self::dfs(
                    i as i32,
                    j as i32,
                    &board,
                    &mut visited,
                    &mut current,
                    &root,
                    &mut result,
                );
            }
        }

        result.into_iter().collect()
    }
}
// @lc code=end
